### PR TITLE
Gdb test script

### PR DESCRIPTION
**Pull request by c1728p9**
This PR adds a script that automatically runs common gdb commands, so gdb regressions can more easily be caught.

### CODE REVIEW

**Comment by c1728p9** [comment]
Verified all tests pass:
```
------ TEST RESULTS ------
Target         Test           Result         Time

kl46z          Basic Test     Pass           4.15199995041
kl46z          Speed Test     Pass           15.4370000362
kl46z          Cortex Test    Pass           3.77399992943
kl46z          Flash Test     Pass           60.3959999084
kl46z          Gdb Test       Pass           117.036000013
lpc1768        Basic Test     Pass           35.6619999409
lpc1768        Speed Test     Pass           35.4450001717
lpc1768        Cortex Test    Pass           6.61599993706
lpc1768        Flash Test     Pass           179.825000048
lpc1768        Gdb Test       Pass           146.566999912
nrf51          Basic Test     Pass           4.49900007248
nrf51          Speed Test     Pass           40.1610000134
nrf51          Cortex Test    Pass           6.28299999237
nrf51          Flash Test     Pass           117.898000002
nrf51          Gdb Test       Pass           131.515000105
k22f           Basic Test     Pass           3.85400009155
k22f           Speed Test     Pass           24.2319998741
k22f           Cortex Test    Pass           4.79200005531
k22f           Flash Test     Pass           77.5309998989
k22f           Gdb Test       Pass           151.927000046
lpc800         Basic Test     Pass           4.53999996185
lpc800         Speed Test     Pass           10.0320000648
lpc800         Cortex Test    Pass           3.25300002098
lpc800         Flash Test     Pass           14.3440001011
lpc800         Gdb Test       Pass           100.046000004
lpc11u24       Basic Test     Pass           7.89599990845
lpc11u24       Speed Test     Pass           9.91699981689
lpc11u24       Cortex Test    Pass           5.04800009727
lpc11u24       Flash Test     Pass           22.9500000477
lpc11u24       Gdb Test       Pass           135.092000008

Test Time: 1480.72900009
All tests passed
```

**Comment by c1728p9** [comment]
Let me know if anyone has feedback on this pull request.  @flit , this test, or  another test very similar could be used to test semi hosting on actual hardware.  Think that would be worth adding in another pull request?

File: pyOCD/tools/gdb_server.py [hunk 1]
```diff
@@ -44,6 +44,10 @@ def __init__(self):
         self.args = None
         self.gdb_server_settings = None
         self.echo_msg = None
+        self.input_args = None
```
Comment (0xc0170) [inline pyOCD/tools/gdb_server.py]: What's the intention here with input_args? The only use I see it's in the test script. How this differ from self.args

File: test/gdb_test.py [hunk 2]
```diff
@@ -15,27 +15,138 @@
...
+# Note
+#  To run this script GNU Tools ARM Embedded must be installed,
+#  along with python for the same architecture.  The program
+#  "arm-none-eabi-gdb-py.exe" requires python for the same
+#  architecture (x86 or 64) to work correctly. Also, on windows
+#  the GNU Tools ARM Embedded bin directory needs to be added to
+#  your path.
 
-from pyOCD.gdbserver import GDBServer
+import os
```
Comment (0xc0170) [inline test/gdb_test.py]: Shall we separete by a blank space local vs external modules

File: test/gdb_test_program/readme.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+How to build gdb test program (windows only right now):
+-Install prerequisites - GNU Tools ARM Embedded 4.9 2014q4 or newer
+-Add bin directory to path - ex. "C:\Program Files (x86)\GNU Tools ARM Embedded\4.9 2015q2\bin" to path
+-Run build.bat to create gdb_test.bin and gdb_test.elf
+-Copy main.bin and main.elf into the test folder
```
Comment (0xc0170) [inline test/gdb_test_program/readme.txt]: The files are in test/gdb_program_test, but generated executables (test/gdb_test.bin/elf) are in the test folder. Shall they be also in the at least same parent folder?

File: test/gdb_script.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+    "init",
+    "reset",
+    "reset halt",
+    "reset halt resume",
+    "halt",
+    "resume",
+    "reg",
+    "arm semihosting enable",
+    "arm semihosting disable",
+    "fawehfawoefhad"
```
Comment (0xc0170) [inline test/gdb_script.py]: what's the use of this magic command - fawehfawoefhad?

File: test/gdb_script.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+                print("  Watchpoint was hit %s" % bkpt_hit)
+                print("  Watchpoint should be hit %s" % should_break)
+                print("  bkpt_size %s, bkpt_access %s, bkpt_address 0x%x, "
+                      "size %s, access %s, addr 0x%x" %
+                      (bkpt_size, bkpt_access, bkpt_addr, size, access, addr))
+                print()
+
+            breakpoint.delete()
+
+        # TODO,future - test reading/writing registers
```
Comment (0xc0170) [inline test/gdb_script.py]: TODO,future? Shall we use just TODO ? I would add a name to identify who can provide more context to this todo. Git blame might not always work (moving files, etc).

**Comment by 0xc0170** [comment]
Can you please rebase?

File: pyOCD/tools/gdb_server.py (hunk 1 above)
Comment (c1728p9) [inline pyOCD/tools/gdb_server.py]: The variable self.args is for passing arguments to the GDBServer class, while input_args are used to override the args passed on the command line.  I'll modify how this is handled to make it more clear.

File: test/gdb_test.py (hunk 2 above)
Comment (c1728p9) [inline test/gdb_test.py]: Done

File: test/gdb_test_program/readme.txt (hunk 3 above)
Comment (c1728p9) [inline test/gdb_test_program/readme.txt]: I modified the gdb test script to use the files in the gdb_test_program folder

File: test/gdb_script.py (hunk 4 above)
Comment (c1728p9) [inline test/gdb_script.py]: It is to make sure the GDB server doesn't crash when an invalid command is sent.  I added a comment to the code to indicate this.

File: test/gdb_script.py (hunk 5 above)
Comment (c1728p9) [inline test/gdb_script.py]: Changed TODO,future to TODO,c1728p9

**Comment by c1728p9** [comment]
Comments addressed and patch rebased.

File: test/gdb_script.py (hunk 5 above)
Comment (0xc0170) [inline test/gdb_script.py]: Often it's TODO (name) or just name ,what do u think?

**Pull request merged**